Add glossary (#5935)
RadeonOpenCompute/ROCm: docs/reference/glossary/device-hardware.rst

.. meta::
  :description: Device hardware glossary for AMD GPUs
  :keywords: AMD, ROCm, GPU, device hardware, compute units, cores, MFMA,
    architecture, register file, cache, HBM

.. _glossary-device-hardware:

***************
Device hardware
***************

This section provides brief definitions of hardware components and architectural
features of AMD GPUs.

.. glossary::
    :sorted:

    AMD device architecture
        AMD's device architecture is based on unified, programmable compute
        engines called Compute Units. See :ref:`hip:hardware_implementation` for
        details.

    Compute units
        Compute Units (CUs) are the fundamental programmable execution engines
        in AMD GPUs that manage thousands of lightweight threads. See

>         :ref:`hip:compute_unit` for details.

Reviewer: This reference shows up as **hip:gfx_ip** in the RTD version? Also, most of the **:ref:** links don't seem to work, except the first one under AMD device architecture, and a few toward the end.

        details.

    Data movement engine
        Data Movement Engines (DMEs) are specialized hardware units in CDNA3 and
        CDNA4 that accelerate multi-dimensional tensor data copies between
        global memory and on-chip memory. See :ref:`hip:dme` for details.

    Compute unit versioning
        Compute Units are versioned with GFX IP identifiers that define their
        microarchitectural features and instruction set compatibility. See
        :ref:`hip:gfx_ip` for details.

    Register file
        The register file is the primary on-chip memory store in each Compute
        Unit, holding data between arithmetic and memory operations. See
        :ref:`hip:memory_hierarchy` for details.

    L1 data cache
        The L1 Data Cache is the private on-chip memory associated with each
        Compute Unit, providing fast access to recently used data. See
        :ref:`hip:vl1`, :ref:`hip:sl1` and :ref:`hip:memory_coherence` for
        details.

    GPU RAM and HBM
        GPU RAM is the large, high-capacity High Bandwidth Memory (HBM)
        subsystem accessible by all Compute Units, forming the foundation of the

>         device's data storage hierarchy. See :ref:`hip:hbm` for details.

Reviewer: Also known as global memory

Reviewer 2: Fixed.

.. meta::
  :description: Device hardware glossary for AMD GPUs
  :keywords: AMD, ROCm, GPU, device hardware, compute units, cores, MFMA,
    architecture, register file, cache, HBM

.. _glossary-device-hardware:

***************

> Device hardware

Reviewer 3: ```suggestion
Device hardware glossary
```

***************

> 

Reviewer 3: How do these definitions line up with the definitions here:
https://rocm.docs.amd.com/en/latest/reference/gpu-arch-specs.html

<img width="1549" height="1517" alt="Image" src="https://github.com/user-attachments/assets/1d664c98-b05b-46ef-8242-423068bf4376" />

Reviewer 3: By the way, should we ensure that all glossary items on that other page reside here?

Author: Good point, I wasn't aware of that page. I'll integrate and update those definitions and place them here.

> This section provides brief definitions of hardware components and architectural

Reviewer 3: This says "brief" - main page says "concise"
Should be consistent

features of AMD GPUs.

.. glossary::
    :sorted:

    AMD device architecture
        AMD's device architecture is based on unified, programmable compute

>         engines called Compute Units. See :ref:`hip:hardware_implementation` for

Reviewer 4: ```suggestion
        engines known as Compute Units (CUs). See :ref:`hip:hardware_implementation` for
```

>         details.

Reviewer 3: ```suggestion
    AMD device architecture
        AMD device architecture is based on unified, programmable compute
        engines known as Compute Units (CUs). See :ref:`hip:hardware_implementation` for
        details.
```

Reviewer 3: Avoid making AMD possessive.

Reviewer 3: Perhaps CUs should link to the CU in the Glossary

    Compute units

>         Compute Units (CUs) are the fundamental programmable execution engines

Reviewer 4: ```suggestion
        CUs are the fundamental programmable execution engines
```

>         in AMD GPUs that manage thousands of lightweight threads. See

Reviewer 3: do you need "thousands of" in here to define compute units?

        :ref:`hip:compute_unit` for details.

    Vector arithmetic logic units
        Vector arithmetic logic units (VALUs) are the primary arithmetic engines
        that execute mathematical and logical operations within AMD Compute
        Units. See :ref:`hip:valu` for details.

    Special function unit
        Special Function Units (SFUs) accelerate transcendental and reciprocal
        mathematical functions such as ``exp``, ``log``, ``sin``, and ``cos``.
        See :ref:`hip:sfu` for details.

>     Load and store unit

Reviewer 3: Is it a slash or "and" - should be consistent:

<img width="1287" height="174" alt="Image" src="https://github.com/user-attachments/assets/e4118058-d342-4f49-a30b-d674a7bf307e" />

>         Load/Store Units (LSUs) handle data transfer between Compute Units and

Reviewer 4: ```suggestion
        Load/Store Units (LSUs) handle data transfer between CUs and
```

        the GPU's memory subsystems, managing thousands of concurrent memory
        operations. See :ref:`hip:lsu` for details.

    Wavefront scheduler
        The Wavefront Scheduler in each Compute Unit decides which group of
        threads to execute each clock cycle, enabling rapid context switching

>         for latency hiding. See :ref:`hip:wave-scheduling` for details.

Reviewer 3: ```suggestion
    Wavefront scheduler
        The wavefront scheduler in each compute unit determines which thread group to execute each clock cycle, enabling rapid context switching
        for latency hiding. See :ref:`hip:wave-scheduling` for details.
```

    SIMD core

>         SIMD Cores are execution lanes that perform scalar and vector arithmetic

Reviewer 3: ```suggestion
        SIMD cores are execution lanes that perform scalar and vector arithmetic
```

>         operations inside each Compute Unit. See :ref:`hip:cdna_architecture`

Reviewer 3: ```suggestion
        operations inside each compute unit. See :ref:`hip:cdna_architecture`
```

Reviewer 3: Why was CU using initial capital letter?

        and :ref:`hip:rdna_architecture` for details.

    Matrix core and MFMA
        Matrix Cores (MFMA units) are specialized execution units that perform
        large-scale matrix operations in a single instruction, delivering high
        throughput for AI and HPC workloads. See :ref:`hip:mfma_units` for

>         details.

Reviewer 3: ```suggestion
    Matrix core (MFMA units) 
        Matrix cores (MFMA units) are specialized execution units that perform
        large-scale matrix operations in a single instruction, delivering high
        throughput for AI and HPC workloads. See :ref:`hip:mfma_units` for
        details.
```

    Data movement engine
        Data Movement Engines (DMEs) are specialized hardware units in CDNA3 and
        CDNA4 that accelerate multi-dimensional tensor data copies between

>         global memory and on-chip memory. See :ref:`hip:dme` for details.

Reviewer 3: ```suggestion
    Data movement engine
        Data Movement Engines (DMEs) are specialized hardware units in CDNA3 and
        CDNA4 that accelerates multi-dimensional tensor data copies between
        global memory and on-chip memory. See :ref:`hip:dme` for details.
```

Reviewer 3: Do you think anyone will know what CDNA is if they're reviewing this definition?
Perhaps this could be "CDNA3 and CDNA4-based architecture"
Or better yet, get rid of CDNA and replace it with AMD Instinct MI300 and MI350 Series - that is more obvious

    Compute unit versioning

>         CUs are versioned with GFX IP identifiers that define their

Reviewer 3: ```suggestion
        Compute unites (CUs) are versioned with GFX IP identifiers that define their
```

Reviewer 3: is GFX uppercase? 
For example, 

<img width="514" height="728" alt="Image" src="https://github.com/user-attachments/assets/f1006e61-053f-4019-900d-e5e180c9f37a" />

should we discuss they are a prefixes followed by numbers and provide examples?

Author: The usage of `GFX` vs `gfx` is pretty inconsistent across AMD docs. It mostly seems to converge towards "`GFX` when talking about generations (GFX 9, GFX 10, etc.)" and "`gfx` when talking about specific GPU targets (gfx950 etc.)".

>         Compute Units are versioned with GFX IP identifiers that define their

Reviewer 4: ```suggestion
        CUs are versioned with GFX IP identifiers that define their
```

        microarchitectural features and instruction set compatibility. See
        :ref:`hip:gfx_ip` for details.

    Register file
        The register file is the primary on-chip memory store in each Compute
        Unit, holding data between arithmetic and memory operations. See
        :ref:`hip:memory_hierarchy` for details.

>     L1 data cache

Reviewer 3: Don't see this term in the other glossary:

<img width="2224" height="713" alt="Image" src="https://github.com/user-attachments/assets/51e54ffa-8eb9-40ba-8f6f-c5fcb519ec74" />

        The L1 Data Cache is the private on-chip memory associated with each
        Compute Unit, providing fast access to recently used data. See
        :ref:`hip:vl1`, :ref:`hip:sl1` and :ref:`hip:memory_coherence` for
        details.

>     GPU RAM and HBM

Reviewer 3: Odd to have to have two terms as a glossary definition.

        GPU RAM, also known as global memory in the HIP programming model, is
        the large, high-capacity High Bandwidth Memory (HBM) subsystem
        accessible by all Compute Units, forming the foundation of the device's

>         data storage hierarchy. See :ref:hip:hbm for details.

Reviewer 3: ```suggestion
        GPU RAM, also known as global memory in the HIP programming model, is
        the large, high-capacity High Bandwidth Memory (HBM) subsystem
        accessible by all Compute Units, and forms the foundation of the device's
        data storage hierarchy. See :ref:hip:hbm for details.
```
